File: shell/internal/storage/model/treefilter.go
package model

// TreeVisibility mirrors src/shared/domain/tree-filter.ts's treeVisibilitySchema — a set of
// hidden node kinds and hidden paths, not a rule list (P28 D12).
type TreeVisibility struct {
	HiddenKinds []string `json:"hiddenKinds"`
	HiddenPaths []string `json:"hiddenPaths"`
}

// EmptyVisibility mirrors src/shared/domain/tree-filter.ts's EMPTY_VISIBILITY.
func EmptyVisibility() TreeVisibility {
	return TreeVisibility{HiddenKinds: []string{}, HiddenPaths: []string{}}
}

File: packages/shared/domain/tree-filter.ts
import { z } from 'zod';
import { nodeKindSchema } from './tree';

// P28 D10-D13: the tree's persisted filter is a set of exclusions, not a rule list. A path or
// kind absent from both arrays is visible — including one fetched for the first time after this
// was saved — and hiding a container never writes entries for what is beneath it (D13): the
// renderer never walks into a row it does not render.
export const treeVisibilitySchema = /*#__PURE__*/ z.object({
  hiddenKinds: /*#__PURE__*/ z.array(nodeKindSchema),
  /** Encoded node paths, relative to the connection — the same strings `TreeNode.path` and
   *  `rowKey()` already use. */
  hiddenPaths: /*#__PURE__*/ z.array(z.string()),
});
export type TreeVisibility = z.infer<typeof treeVisibilitySchema>;

export const EMPTY_VISIBILITY: TreeVisibility = { hiddenKinds: [], hiddenPaths: [] };
